operators/tokenizer/bpe_kernels.cc

```diff
@@ -660,95 +659,79 @@ struct VectorEqual {
-  }
-
-  // consider to use SAX parser for large json file
-  nlohmann::json tok_json;
-  std::ifstream module_ifs;
-
-  // Following vocab and merges only used for tiktoken case but accessed outside scope below
-  std::unordered_map<std::string, uint32_t> vocab;
-  std::vector<std::pair<std::string, std::string>> merges;
-
-  if (tiktoken_){
-    std::string module_file = config.GetTikTokenModuleFile();
+OrtxStatus JsonFastTokenizer::LoadAddedTokens(const json& tok_json, const ort_extensions::bpe::TokenJsonConfig& config) {
+  auto added_tokens = tok_json.find("added_tokens");
+  if (added_tokens != tok_json.end()) {
+    for (const auto& token : *added_tokens) {
+      bpe::AddedToken added_token;
+      added_token.id_ = token.value("id", 0);
+      added_token.token_type_ = token.value("__type", "");
+      added_token.content_ = token.value("content", "");
+      added_token.lstrip_ = token.value("lstrip", false);
+      added_token.normalized_ = token.value("normalized", false);
+      added_token.rstrip_ = token.value("rstrip", false);
+      added_token.single_word_ = token.value("single_word", false);
+      added_token.special_ = token.value("special", false);
 
-    module_ifs = path(module_file).open();
-    if (!module_ifs.is_open()) {
-      return OrtxStatus(kOrtxErrorInvalidFile, "Failed to open module file: " + module_file);
+      added_tokens_.emplace_back(added_token);
+      if (added_token.content_ == config.bos_token_) {
+        bos_token_id_ = added_token.id_;
+      } else if (added_token.content_ == config.eos_token_) {
+        eos_token_id_ = added_token.id_;
+      } else if (added_token.content_ == config.pad_token_) {
+        pad_token_id_ = added_token.id_;
+      }
     }
+  }
 
-    std::unordered_map<std::vector<uint8_t>, uint32_t, VectorHash, VectorEqual> bpe_ranks;
-
-    std::string line;
-    while (std::getline(ifs, line)) {
-      if (!line.empty()) {
-        std::istringstream lineStream(line);
-        std::string token;
-        uint32_t rank;
-        while (lineStream >> token >> rank) {
-            // Decode base64 token and convert rank to int
-            std::vector<uint8_t> decoded_token;
-            base64_decode(token, decoded_token);
-            // Store bpe token and rank
-            bpe_ranks[decoded_token] = rank;
+  return bbpe_tokenizer_->LoadAddedTokens(added_tokens_);
+}
+
+// Helper methods (to be added to the class declaration)
+bool JsonFastTokenizer::CheckForSpmModel(const json& tok_json) {
+  auto decoder_node = tok_json.find("decoder");
+  if (decoder_node != tok_json.end()) {
+    auto decoders_node = decoder_node->find("decoders");
+    if (decoders_node != decoder_node->end() && decoders_node->is_array()) {
+      for (const auto& step : *decoders_node) {
+        std::string type = step.value("type", "");
+        if (type == "Replace") {
+          std::string target = step.value("/pattern/String"_json_pointer, "");
+          if (target == "\xe2\x96\x81") {
```

Comment:
can we also parse this value from somewhere (like the module file) rather than hardcoding it?

Reply:
yes, define a constant variable and removed the duplicates.